### PR TITLE
GAWB 296: Update all related intersection groups when changing any group

### PR DESCRIPTION
**Pull request by MatthewBemis**
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] Tell  that the PR exists if he wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [ ] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [ ]  sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Squash commits, rebase if necessary
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Merge to develop 
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: Check configuration files in Jenkins in case they need changes
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 1]
```diff
@@ -1338,9 +1338,21 @@ class WorkspaceService(protected val userInfo: UserInfo, dataSource: DataSource,
               containerDAO.authDAO.saveGroup(group, txn)
               level -> group
             }
+
+            def addOwnerToOwnerIntersectionGroup(ownerIntersectRef: RawlsGroupRef) = {
+              val ownerIntersect = containerDAO.authDAO.loadGroup(ownerIntersectRef, txn).getOrElse(throw new RawlsException("Unable to add owner to group"))
+              gcsDAO.addMemberToGoogleGroup(ownerIntersect, Left(RawlsUser(userInfo)))
+              containerDAO.authDAO.saveGroup(ownerIntersect.copy(users = Set(RawlsUser(userInfo))), txn)
+            }
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: 2 problems with this function: 1) it should probably be in the gcs dao setupWorkspace call since that is where the rest of the logic is. 2) addMemberToGoogleGroup returns a future the result of which is lost.

File: src/main/scala/org/broadinstitute/dsde/rawls/RawlsException.scala [hunk 2]
```diff
@@ -5,4 +5,4 @@ import spray.http.{StatusCodes, StatusCode}
 
 class RawlsException(message: String = null, cause: Throwable = null) extends Exception(message, cause)
 
-class RawlsExceptionWithErrorReport(val errorReport: ErrorReport) extends RawlsException(errorReport.message)
\ No newline at end of file
+class RawlsExceptionWithErrorReport(val errorReport: ErrorReport) extends RawlsException(errorReport.toString)
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/RawlsException.scala]: Why this change?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/AuthDAO.scala [hunk 3]
```diff
@@ -18,6 +18,8 @@ trait AuthDAO {
 
   def loadGroupByEmail(groupEmail: String, txn: RawlsTransaction): Option[RawlsGroup]
 
+  def findRelatedWorkspaces(groupRef: RawlsGroupRef, txn: RawlsTransaction): Seq[Workspace]
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/AuthDAO.scala]: Rename to findWorkspacesForRealm or similar

Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/AuthDAO.scala]: Or am I misunderstanding what this does?  Either way, I would like a comment describing it

File: src/main/scala/org/broadinstitute/dsde/rawls/RawlsException.scala (hunk 2 above)
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/rawls/RawlsException.scala]: Error details were getting left out. Unrelated to this PR.

File: src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala [hunk 4]
```diff
@@ -390,6 +370,31 @@ class UserService(protected val userInfo: UserInfo, dataSource: DataSource, prot
     }
   }
 
+  def overwriteGroupMembersInternal(group: RawlsGroup, users: Set[RawlsUser], subGroups: Set[RawlsGroup], txn: RawlsTransaction): Future[Option[ErrorReport]] = {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:373]: Any changes from the non-internal version?  Looks identical to me but I wanted to check.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/AuthDAO.scala (hunk 3 above)
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/AuthDAO.scala]: findWorkspacesForGroup is good. I'll also add a comment. It finds all of the workspaces that have the specified group in any part of its ACL tree.

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 5]
```diff
@@ -1364,15 +1376,21 @@ class WorkspaceService(protected val userInfo: UserInfo, dataSource: DataSource,
...
-  private def requireCreateWorkspaceAccess(workspaceName: WorkspaceName, txn: RawlsTransaction)(op: => Future[PerRequestMessage]): Future[PerRequestMessage] = {
-    containerDAO.billingDAO.loadProject(RawlsBillingProjectName(workspaceName.namespace), txn) match {
+  private def requireCreateWorkspaceAccess(workspaceRequest: WorkspaceRequest, txn: RawlsTransaction)(op: => Future[PerRequestMessage]): Future[PerRequestMessage] = {
+    containerDAO.billingDAO.loadProject(RawlsBillingProjectName(workspaceRequest.toWorkspaceName.namespace), txn) match {
       case Some(billingProject) =>
         if (billingProject.users.contains(RawlsUser(userInfo))) {
-          op
+          workspaceRequest.realm match {
+            case Some(realm) => containerDAO.authDAO.loadGroupIfMember(realm, RawlsUser(userInfo), txn) match {
+              case None => Future.failed(new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.Forbidden, "You cannot create a workspace in a realm that you do not have access to")))
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: Give the name of the realm in the error

File: src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala (hunk 4 above)
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:373]: No changes

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala [hunk 6]
```diff
@@ -551,6 +553,111 @@ class WorkspaceApiServiceSpec extends ApiServiceSpec {
       }
   }
 
+  it should "return 403 when creating a workspace in a realm that you don't have access to" in withTestDataApiServices { services =>
+    val realmGroup = RawlsGroup(RawlsGroupName("realm-for-testing"), RawlsGroupEmail("king@realm.example.com"), Set.empty, Set.empty)
+
+    services.gcsDAO.adminList += testData.userOwner.userEmail.value
+
+    Post(s"/admin/groups", realmGroup) ~>
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala:561]: Is there a specific reason for doing this instead of containerDAO.authDAO.saveGroup ?

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala [hunk 7]
```diff
@@ -551,6 +553,111 @@ class WorkspaceApiServiceSpec extends ApiServiceSpec {
...
+
+    Post(s"/workspaces", httpJson(workspaceWithRealm)) ~>
+      sealRoute(services.workspaceRoutes) ~>
+      check {
+        assertResult(StatusCodes.Created) {
+          status
+        }
+      }
+
+    //add userWriter to both workspace ACLs
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala]: Comment doesn't match action: userWriter is added to Read and userOwner is added to Owner

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala (hunk 6 above)
Comment (MatthewBemis) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala:561]: working with the DAOs directly doesn't create the mock google group behind the rawls group

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala (hunk 7 above)
Comment (MatthewBemis) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala]: oops this test originally had two workspaces in it before i changed how it worked. will make sure all comments still make sense

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala [hunk 8]
```diff
@@ -551,6 +553,111 @@ class WorkspaceApiServiceSpec extends ApiServiceSpec {
...
+    Patch(s"/workspaces/${workspaceWithRealm.namespace}/${workspaceWithRealm.name}/acl", httpJson(Seq(WorkspaceACLUpdate(testData.userWriter.userEmail.value, WorkspaceAccessLevels.Read), WorkspaceACLUpdate(testData.userOwner.userEmail.value, WorkspaceAccessLevels.Owner)))) ~>
+      sealRoute(services.workspaceRoutes) ~>
+      check {
+        assertResult(StatusCodes.OK) { status }
+      }
+
+    //assert userWriter is not a part of realmReaderACLs
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala]: show that userOwner is a member of the Owner ACL, also

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala (hunk 7 above)
Comment (MatthewBemis) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala]: on second thought i'll also have userWriter added to the writer group instead of the reader group because that's counterintuitive

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala [hunk 9]
```diff
@@ -551,6 +553,111 @@ class WorkspaceApiServiceSpec extends ApiServiceSpec {
...
+      sealRoute(services.workspaceRoutes) ~>
+      check {
+        assertResult(StatusCodes.Forbidden) {
+          status
+        }
+      }
+
+  }
+
+  it should "update the intersection groups for related workspaces when group membership changes" in withTestDataApiServices { services =>
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala:586]: Need additional tests like this one, or add steps to this test:
- remove user from Realm -> user is no longer on ACL
- group A is a member of group B, which is a member of a Workspace ACL.  add and remove user from group A -> check that ACLs are updated appropriately
- group C is a member of group D, which is a member of a Realm.  add and remove user from group C -> check that ACLs are updated appropriately

File: src/main/scala/org/broadinstitute/dsde/rawls/RawlsException.scala (hunk 2 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/RawlsException.scala]: ok

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala (hunk 6 above)
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala:561]: is that ever used here, though?

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala (hunk 7 above)
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala]: yes :)

**Comment by jmthibault79** [comment]
Done with review

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala (hunk 6 above)
Comment (MatthewBemis) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala:561]: not directly, but it caused errors behind the scenes if it wasn't there

Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala:561]: ok

**Comment by jmthibault79** [comment]
A few minor comments then :+1:
